src/elfcore.cc

```diff
@@ -711,30 +721,151 @@ static Ehdr *SanitizeVDSO(Ehdr *ehdr, size_t start, size_t end) {
   return ehdr;
 }
 
+static int local_atoi(const char *s) {
+  int n = 0;
+  int neg = *s == '-';
+  if (neg) s++;
+  while (*s >= '0' && *s <= '9') n = 10 * n + (*s++ - '0');
+  return neg ? -n : n;
+}
+
+/* This function will update the prinfo of the process based on the key
+ * and value passed.
+ */
+static void UpdatePsInfo(prpsinfo *prinfo, char *key, char *value) {
+  if (strncmp(key, "Uid", 3) == 0) {
+    /* Only Uid and Gid will have 4 fields and separted by tabs */
+    char *split_ptr = strchr(value, '\t');
+    *split_ptr = '\0';
+    prinfo->pr_uid = local_atoi(value);
+  } else if (strncmp(key, "Gid", 3) == 0) {
+    char *split_ptr = strchr(value, '\t');
+    *split_ptr = '\0';
+    prinfo->pr_gid = local_atoi(value);
+  } else if (strncmp(key, "PPid", 4) == 0) {
+    prinfo->pr_ppid = local_atoi(value);
+  } else if (strncmp(key, "NSpgid", 6) == 0) {
+    prinfo->pr_pgrp = local_atoi(value);
+  }
+}
+
+/*
+ * This function reads the /proc/[pid]/status file to find the prinfo structure
+ */
+static void ReadPsStatus(prpsinfo *prinfo, pid_t pid) {
+  struct io io_status;
+  char proc_path[80];
+  char *key_start, *value_start;
+  char *ind;
+
+  prinfo->pr_pid = pid;
+  prinfo->pr_sid = sys_getsid(pid);
+  prinfo->pr_sname = 'R';  // Emulating the process as Running
+
+  io_status.data = io_status.end = 0;
+  sprintf(proc_path, "/proc/%d/status", pid);
+  NO_INTR(io_status.fd = sys_open(proc_path, O_RDONLY, 0));
+  if (io_status.fd >= 0) {
+    if (GetChar(&io_status) > 0) {
+      ind = (char *)io_status.buf;
+      key_start = ind;
+      while (ind < (char *)io_status.end) {
+        if (*ind == ':') {
+          *ind = '\0';
+          value_start = ind + 2; /* To skip the :<tab> */
+        }
+        if (*ind == '\n') {
+          *ind = '\0';
+          UpdatePsInfo(prinfo, key_start, value_start);
+          key_start = ++ind;
+        }
+        ind++;
+      }
+    }
+  }
+}
+
+/*
+ * This function helps to find, if we can access the given
+ * address from a process memory particularly vdso section.
+ */
+static int
+PeekVDSO(void *addr, void *dest, size_t size, pid_t pid)
```

Comment:
given there is nothing vdso specific in this function, it should have a more generic name

Reply:
Done, changed the function name.